pgxn/neon/relkind_cache.c

```diff
@@ -0,0 +1,309 @@
+/*-------------------------------------------------------------------------
+ *
+ * relkind_cache.c
+ *      Cache to track the relkind of relations
+ *
+ * Portions Copyright (c) 1996-2021, PostgreSQL Global Development Group
+ * Portions Copyright (c) 1994, Regents of the University of California
+ *
+ *-------------------------------------------------------------------------
+ */
+#include "postgres.h"
+
+#include "neon.h"
+#include "miscadmin.h"
+#include "neon_pgversioncompat.h"
+
+#include "pagestore_client.h"
+#include RELFILEINFO_HDR
+#include "storage/smgr.h"
+#include "storage/lwlock.h"
+#include "storage/ipc.h"
+#include "storage/shmem.h"
+#include "catalog/pg_tablespace_d.h"
+#include "utils/dynahash.h"
+#include "utils/guc.h"
+
+#if PG_VERSION_NUM >= 150000
+#include "miscadmin.h"
+#endif
+
+/*
+ * The main goal of this cache is to avoid calls of mdexists in neon_write,
+ * which is needed to distinguish unlogged relations.
+ *
+ * This hash is also used to mark relation during unlogged build.
+ * It has limited size, implementing eviction based on LRU algorithm.
+ * Relations involved in unlogged build are pinned in the cache (assuming that
+ * number of concurrent unlogged build is small.
+ *
+ * Another task of this hash is to prevent race condition during unlogged build termination.
+ * Some backend may want to evict page which backenf performing unlogged build can complete it and unlinking local files.
+ * We are using shared lock which is hold during all write operation. As far as lock is shared is doesn't prevent concurrent writes.
+ * Exclusive lock is taken by unlogged_build_end to change relation kind.
+ */
+
+typedef struct
+{
+	size_t      size;
+	uint64		hits;
+	uint64		misses;
+	uint64		pinned;
+	slock_t		mutex;
```

Comment:
How about using a LWLock? spinlock seems a little too low-level, we usually use LWLocks to protect hash tables like this.

Reply:
I think that LWLock is mostly used when we need to obtain shared and exclusive locks. 
For relkind cache shared lock is never used because of LRU list ()as well as in relsize cache).
Should I still  prefet LWLock to spinlock?
Yes, spinlock seems to be more lowlevel but has less overhead.
relkind cache is not expected to be locked for a long time...

Reply:
Yeah, it's not just about having shared and exclusive locks. The similar lock in relsize_cache.c is also only used in exclusive mode, for example. There's little downside to using a LWLock, it's very fast when there's no contention.

"for a long time" is relative :-).

Reply:
Changed